File: Leetcode First 200 JS solution/1. Two Sum.js

var twoSum = function(nums, target){
    let hashmap = new Map();
    let ans = [];

    for (let i = 0; i < nums.length; i++) {
        if (hashmap.has(target - nums[i])) {
            // object property with a specific key is much faster than creating
            // a Map or using get
            ans.push(...[i, hashmap.get(target - nums[i])]);
            break;
        }
        hashmap.set(nums[i], i);
    }
    return ans;
}

// time complexity o(n)
// space complexity o(n)

// using object

var twoSum = function(nums, target) {
    let hashmap = {};
    for (let i = 0; i < nums.length; i++) {
        const key = target - nums[i];
        if (key in hashmap) {
            return [i, hashmap[key]];
        }
        hashmap[nums[i]] = i;
    }
}

File: Leetcode First 200 C++ solution/1.two sum.cpp
#include <vector>
#include <unordered_map>
using namespace std;

class Solution{
    public:
        vector<int> twoSum(vector<int>& nums, int target) {
            unordered_map<int, int> heap;
            for (int i = 0; i < nums.size(); i++) {
                int r = target - nums[i];
                if (heap.count(r)) return {heap[r], i};
                heap[nums[i]] = i;
            }
            return {};
        }
};


